README

Canna 3.7p3 をベースに fork した。https://www.nslabs.jp/patch.rhtml
目標:
 - `canuum/` を削除  -- 伴って, ライセンスから GPL を削除. [済み]
 - `wchar_t` は環境によって幅が違ってポータブルではない。`_WCHAR_T`, `CANNA_NEW_WCHAR_AWARE`, `CANNA_WCHAR16` は無視して, `uint16_t` 決め打ち. 内部コードはEUC-JP. [済み]
 - `lib/RKindep/cksum.c` に4条項BSDライセンス (宣伝条項付き) が適用されていた。ファイルを FreeBSD 由来に入れ替え, 3条項BSDライセンス (MITライセンスと同等) に変更. [済み]
 - `ENGINE_SWITCH` マクロ, `lib/canna/engine.c` 削除. See `misc/engine.cpp`. cannaserver or irohaserver / jserver (Wnn4) / sj3serv (SJ3) の切り替え。すでに正常に動いていない。
 - imake から autoconf & automake に変更.
 - Windows にポーティング




              CANNA Version 3.6, a Japanese input system

                             Canna Project
		https://ja.osdn.net/projects/canna/

0. Preface

       This file describes the introduction of Canna.  For
  installation of Canna or changes from the previous versions of
  Canna, see the file `INSTALL' or `CHANGES.jp' respectively. (Sorry,
  but `INSTALL' and `README' may not be updated compared to the files
  written in Japanese.  Also sorry that `CHANGES' is provided only in
  Japanese.)

1. Overview

       This directory includes files which are source codes for a
  Japanese input system named Canna.  Canna provides a unified user
  interface for Japanese input.

  *Note:  Canna was called Iroha among developers of Canna
          previously. 

       There exists a commonly usable Japanese input system called
  Wnn.  Canna gives an alternative Japanese input system.

       Canna is based on a client-server model for kana-to-kanji
  conversion, that is, an application program which uses a Japanese
  input system communicates with a kana-to-kanji conversion server,
  which is a separate process from the application program.  Canna has
  several features as follows:

  1) Based on client-server model kana-to-kanji conversion
  2) Supporting automatic kana-to-kanji conversion
  3) Providing a unified user interface to input Japanese
  4) Supporting customization by users
  5) Based on lisp language customization
  6) A mechanism to add a suitable part of speech to each registered
     words by users
  7) Providing a library which supports a unified user interface
  8) Maintenance tools for conversion dictionaries
  9) A customizing tool which supports easy customization
 10) Supporting Nemacs(Mule), kinput2 and uum

       Below, we would like to describe details of Canna's features.

2. Features of Canna

2.1 Based on client-server model kana-to-kanji conversion

       Canna converts kana to kanji based on a client-server model.
  That is, an application program communicates with a kana-to-kanji
  conversion server to achieve Japanese input.

2.2 Supporting automatic kana-to-kanji conversion

       Canna is the first free Japanese input system which supports
  automatic kana-to-kanji conversion which based on client-server
  model.  Generally, it is not easy to operate Japanese with automatic
  kana-to-kanji conversion in PC environment.  However, Canna makes it
  easier for supporting better user interface which extended their
  generally(unautomatically) conversion.

2.3 Providing a unified user interface to input Japanese

       Developers of Canna used to use Egg on Nemacs when we had to
  input Japanese.  Egg was very harmonious with Nemacs and its user
  interface was convenient.

  *Note: Nemacs is a Japanese version of Emacs.  Egg is an interface
         between Nemacs and Wnn.  Egg provides a user interface by
         itself.

       On the other hand, if we had to input Japanese without Nemacs,
  for example, to input Japanese in command line of shell, we were not
  able to use Egg because Egg was usable only in Emacs environment.
  On the other hand, there is also a Japanese input system for TTYs
  called Uum, Uum has a different user interface from Egg and we did
  not become accustomed with it.

       Especially in using X clients, we were eager to use Egg's user
  interface to use them.  We found a lot of good tools on the X window
  system such as xmh and xcalendar.  Our desire was to localize those
  clients into our national language and to operate them with Egg's
  user interface.

<<Let's create another Egg by ourselves>>

       Thus, we decided to create an Egg like system and to provide it
  in a library.  This is the Canna.  Now, Canna provides more features
  than Egg.  Canna can be used in Emacs, in X environments, and also
  on TTYs.

2.4 Supporting customization by users

       As well as key binding, it is possible to customize
  romaji-to-kana conversion rules, status describing strings,
  dictionaries, etc.  A customization is described in a customization
  file.  A customization file can be shared among applications using
  Canna.

2.5 Based on lisp language customization

       Canna used to use a simple syntax for a customization file.
  But once a new function is introduced to the syntax of the
  customization file, old application programs will have to be
  re-written to incorporate the new rules and conditions of Canna.

       According to use new customization syntax based on lisp
  language, it is possible to ignore the descriptions which old
  version's Canna cannot understand.  With new syntax, you can get
  version of Canna or connected server, and it is also possible to
  customize on conditions of these versions.

2.6 A mechanism to add a suitable part of speech to each registered
    words by users

       When we made our own user interface, we added one new idea into
  the word registration part of Canna.

       In kana-to-kanji conversion system, more detailed parts of
  speech are used than what are used in school grammars.  Canna uses
  almost 400 parts of speech.  On the other hand, it might be almost
  impossible and unkind to ask users what kind of a part of speech
  must be used for the registered word.

       In Canna, we made a new mechanism to add a suitable part of
  speech to registered words.  Canna shows several sentences using
  newly registered word, and ask user the usage is correct or not.
  Asking several questions to user helps Canna to decide what part of
  speech is suitable for the newly registered word.

2.7 Providing a library which supports a unified user interface

       A user-interface library is provided.  With this library,
  programmers can easily add a Japanese input system with unified user
  interface to application programs.

       This library has higher-level functions than those Wnn's
  high-level library has, and interpretation of each input key is also
  processed inside of this library.

       An interface to this library is simple.  Giving input key to
  this library returns several pieces of information to display
  pre-edit status.  Interpretations of key functions are hidden inside
  of this library, and it is not necessary for application programs to
  be conscious of them.

       This library meets the case of treating inputs from several
  windows.  By giving context identifiers which correspond to each
  window with key inputs to this library makes it possible to process
  plural kana-to-kanji conversion concurrently in a process.

2.8 Maintenance tools for conversion dictionaries

       Canna provides more than ten tools to maintain kana-to-kanji
  conversion dictionaries.  Using these tools, you can do the
  following things.

   - Conversion between text format dictionaries and binary format
     ones.
   - Creating, deleting, listing, renaming dictionaries by remote
     operation
   - Uploading and downloading dictionaries
   - Adding/deleting items to/from a dictionary by a batch procedure

2.9 A customizing tool which supports easy customization

       In place of editing a customization file directly, it is
  possible to set up customization by using a customizing tool.

2.10 Supporting Nemacs(Mule), kinput2 and uum

       Canna provides a patch for Nemacs(Mule), which makes
  Nemacs(Mule) possible to input Japanese in Canna's unified user
  interface.  And, after Mule 0.9.5, the distribution of Mule includes
  interface to Canna.

       If kinput2 use Canna as kana-to-kanji conversion engine,
  application programs which use kinput can use Canna's unified user
  interface to input Japanese

       Additionally, Canna supports a patch for uum, which makes
  possible to input Japanese in Canna on TTYs.  For uum, see
  canuum/README.jp.  (Sorry, but canuum/README.jp is in Japanese.)

3.  Contents of Canna

       Canna contains the following things.

   - Kana-to-kanji conversion server (cannaserver)
   - User interface library (libcanna.a, libcanna.so.1)
   - Remote version of dictionary accessing library (libRKC.a)
   - Local version of dictionary accessing library (libRK.a)
   - Maintenance tools for conversion dictionaries
   - Patch files for uum of Canna version
   - A sample program

4. Documents

       Under this directory, there are several documents.  Most of
  them are in Japanese.

  1) Document of Canna (in Japanese)

       The document is in jlatex format and is in doc/man/guide.
  This document describes the general comments of how to use Canna.

  2) Documents for commands

       The documents are in man macro format of roff and are in the
  respective directory.  Source codes are also available in same
  directory.  Japanese documents have suffix .jmn, and English have
  .man.

  3) Documents for application interface library

       The documents are in man macro format of roff and are in
  lib/canna and lib/RK.

  4) Document for kana-to-kanji conversion protocol (in Japanese)

       The document is in jlatex format and is in doc/intern.  Canna
  has the protocol versions 1.0, 1.1, 2.0, 2.1, 3.0 and 3.2.  The
  document of the protocol version 1.0 and 1.1 is the file
  proto12.tex.  Other version's document is protocol.tex.

  5) Document for lisp language of customization file (in Japanese)

       The document is in jlatex format and is in doc/lisp as
  canlisp.tex.

  6) Document for uum of Canna version

       The notes is in text file format and is in canuum as README.jp.

  7) Other Documents

       The documents are in jlatex format and are in doc/misc.

5. Bugs

       Canna have some bugs as follows:

  (1) Canuum doesn't run on Solaris 2.1.

  (2) Canuum doesn't run on SunOS 4.2 with compiler gcc.

  (3) On SONY NEWS, the terminal mode on pseudo tty is different which
      make by canuum.

  (4) No English documents as follows:
        canuum/canuum.man

  (5) When using automatic kana-to-kanji conversion, you sometimes can
      not convert correctly.

  (6) On Solaris 2.3, you will have some cores.

  (7) When you compile, you will have a warning message "Undefined row
      vectors: BM".  Don't worry it.



debian/changelog
canna (3.7p3-19) unstable; urgency=medium

  * QA upload.

  [ Debian Janitor ]
  * Drop fields with obsolete URLs.
  * Fix day-of-week for changelog entry 32p2+uf1+gp3-0.

 -- Jelmer Vernooĳ <jelmer@debian.org>  Fri, 27 May 2022 17:59:53 +0100

canna (3.7p3-18) unstable; urgency=medium

  * QA upload.
  * debian/rules: Set FORCE_SOURCE_DATE=1 in order for texlive to respect
    SOURCE_DATE_EPOCH when generating .dvi file. (Closes: #995401)
  * libcann1g: Remove explicit calls to ldconfig from postinst and postrm,
    now handled by dpkg triggers.
  * Update to debhelper-compat 13.
  * canna: Avoid recursive chown to /var/lib/canna and /var/log/canna by
    explicitly calling the files to be chowned.

 -- Vagrant Cascadian <vagrant@debian.org>  Mon, 09 May 2022 13:25:00 -0700

canna (3.7p3-17) unstable; urgency=high

  * QA upload.

  [ Jenkins ]
  * Remove constraints unnecessary since stretch.

  [ Boyuan Yang ]
  * debian/control:
    + Remove useless build-dep autoconf.
    + Bump Standards-Version to 4.6.0.
  * debian/rules:
    + Split dh_installman invocations to debian/*.manpages files.
    + Do not call dh_auto_clean in dh_auto_build target.
    + Enable verbose build.
    + Use buildflags.mk to provide build flags.
  * debian/autoreconf: Specify autoreconf directories.
  * debian/dirs: Rename to debian/canna.dirs to avoid confusion.
  * debian/patches/13_fix-ar-clq.patch: Workaround "ar clq" problem with
    binutils 2.37+ (see #981072). (Closes: #978778)
  * debian/postinst: Use mktemp instead of tempfile (Closes: #992455).

 -- Boyuan Yang <byang@debian.org>  Tue, 14 Sep 2021 02:18:59 -0400

canna (3.7p3-16) unstable; urgency=medium

  * QA upload.
  * Configured git-buildpackage for Debian.
  * Ran wrap-and-sort -a.
  * debian/canna.lintian-overrides:
      - Removed unused override 'spelling-error-in-binary'.
      - Renamed tag binary-without-manpage to no-manual-page.
  * debian/canna-utils.lintian-overrides: renamed tag binary-without-manpage to
    no-manual-page.
  * debian/control: added 'Pre-Depends: ${misc:Pre-Depends}' as per lintian
    warning.
  * debian/libcanna1g.symbols: added 'Build-Depends-Package' field.
  * debian/patches:
      - 05_hppabuild.patch: removed patch applied by upstream.
      - 06_fix_spelling_error.patch renamed to 05_fix_spelling_error.patch.
      - 07_fix_manpages_error.patch renamed to 06_fix_manpages_error.patch.
      - 08_fix_ftbfs_on_hurd-i386.patch renamed to
        07_fix_ftbfs_on_hurd-i386.patch.
      - 09_fix_format_security.patch renamed to 08_fix_format_security.patch.
      - 10_fix_manpages_error.patch renamed to 09_fix_manpages_error.patch.
      - 11_fix_configure.ac.patch renamed to 10_fix_configure.ac.patch.
      - 11_fix_spelling_error_in_binary.patch: added.
      - 12_make_the_output_of_mkbindic_reproducible.patch: added.
        Thanks to Chris Lamb <lamby@debian.org>. (Closes: #861955)
  * debian/po/pt_BR.po: converted from ISO-8859-1 to UTF-8.
  * debian/rules:
      - Removed 'override_dh_auto_clean'.
      - Removed trailing whitespaces.

 -- Francisco Vilmar Cardoso Ruviaro <francisco.ruviaro@riseup.net>  Tue, 01 Sep 2020 03:14:05 +0000

canna (3.7p3-15) unstable; urgency=medium

  * QA upload.
  * Using new DH level format. Consequently:
      - debian/compat: removed.
      - debian/control: changed from 'debhelper' to 'debhelper-compat' in
        Build-Depends field and bumped level to 12.
  * debian/control:
      - Added 'Rules-Requires-Root: binary-targets' to source stanza.
      - Bumped Standards-Version to 4.5.0.
      - Removed 'autotools-dev' from Build-Depends field.
      - Updated VCS fields to use Salsa.
      - wrap-and-sort.
  * debian/copyright:
      - Migrated to 1.0 format.
      - Updated packaging copyright.
  * debian/patches/11_fix_configure.ac.patch:
      - Added to fix FTBFS caused after the DH level has bumped.
  * debian/rules:
      - Added DEB_BUILD_MAINT_OPTIONS to improve the GCC hardening.
  * debian/salsa-ci.yml:
      - Added to provide CI tests for Salsa.

 -- Francisco Vilmar Cardoso Ruviaro <francisco.ruviaro@riseup.net>  Tue, 30 Jun 2020 22:10:14 +0000

canna (3.7p3-14) unstable; urgency=medium

  * QA upload.
  * Make cannakill a symlink to catdic, fixing the file conflict
    between canna-dbgsym and canna-utils-dbgsym. (Closes: #868541)
  * Add the missing lsb-base dependency to canna.

 -- Adrian Bunk <bunk@debian.org>  Sun, 16 Jul 2017 16:59:14 +0300

canna (3.7p3-13.1) unstable; urgency=medium

  * Non-maintainer upload to fix FTBFS.
      Use only autoconf without version suffix.

 -- YunQiang Su <syq@debian.org>  Mon, 05 Oct 2015 17:13:17 +0800

canna (3.7p3-13) unstable; urgency=medium

  * QA upload.
  * debian/control: add texlive-lang-japanese to B-D: to fix FTBFS
    (Closes: #750320).
  * support init.d status
    - debian/control: add canna-utils to canna Depends.
    - debian/init: add status.
  * debian/init: support /lib/lsb/init-functions.
  * debian/libcanna1g.lintian-overrides: remove unused-override.
  * bumped Standards-Version to 3.9.5.

 -- HIGUCHI Daisuke (VDR dai) <dai@debian.org>  Wed, 04 Jun 2014 21:35:31 +0900

canna (3.7p3-12) unstable; urgency=low

  * QA upload.
  * enable hardening.
    - debian/rules: add dpkg-buildflags.
    - debian/patches/09_fix_format_security.patch: new file.
  * debian/patches/10_fix_manpages_error.patch: new file.
  * lintian-overrides: binary-without-manpage and spelling-error-in-binary
    - debian/canna-utils.lintian-overrides: new file.
    - debian/canna.lintian-overrides: new file.
    - debian/libcanna1g.lintian-overrides: new file.
  * debian/libcanna1g.symbols: new file.
  * bumped Standards-Version to 3.9.4.

 -- HIGUCHI Daisuke (VDR dai) <dai@debian.org>  Sun, 09 Jun 2013 04:46:26 +0900

canna (3.7p3-11) unstable; urgency=medium

  * QA upload.
  * debian/control:
    - Add Breaks/Replaces (<< 3.7p3-10) to canna-utils, which now owns
      /usr/bin/chmoddic (Closes: #694266).


canna/RK.h
/* Copyright 1993 NEC Corporation, Tokyo, Japan.
 *
 * Permission to use, copy, modify, distribute and sell this software
 * and its documentation for any purpose is hereby granted without
 * fee, provided that the above copyright notice appear in all copies
 * and that both that copyright notice and this permission notice
 * appear in supporting documentation, and that the name of NEC
 * Corporation not be used in advertising or publicity pertaining to
 * distribution of the software without specific, written prior
 * permission.  NEC Corporation makes no representations about the
 * suitability of this software for any purpose.  It is provided "as
 * is" without express or implied warranty.
 *
 * NEC CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
 * NO EVENT SHALL NEC CORPORATION BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
 * USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTUOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

/* $Id: RK.h,v 1.10 2003/09/21 10:16:49 aida_s Exp $ */
#ifndef		_RK_h
#define		_RK_h


#ifdef canna_export
# define CANNA_EXP_PREDEFINED
#else /* canna_export has not been not defined */
# define canna_export(x) x
#endif

#ifndef pro
  #define pro(x) x
#endif

#include <canna/sysdep.h>

typedef	struct {
   int		ylen;		/* yomigana no nagasa (in byte) */
   int		klen;		/* kanji no nagasa (in byte) */
   int		rownum;		/* row number */
   int		colnum;		/* column number */
   int		dicnum;		/* dic number */
}		RkLex;

typedef	struct {
   int		bunnum;		/* bunsetsu bangou */
   int		candnum;	/* kouho bangou */
   int		maxcand;  	/* sou kouho suu */
   int		diccand;	/* jisho ni aru kouho suu */
   int		ylen;		/* yomigana no nagasa (in byte) */
   int		klen;		/* kanji no nagasa (in byte) */
   int		tlen;		/* tango no kosuu */
}		RkStat;

struct DicInfo {
    unsigned char	*di_dic;
    unsigned char	*di_file;
    int			di_kind;
    int			di_form;
    unsigned		di_count;
    int			di_mode;
    long		di_time;
};

/* romaji/kanakanji henkan code */
#define	RK_XFERBITS	4	/* bit-field width */
#define	RK_XFERMASK	((1<<RK_XFERBITS)-1)
#define	RK_NFER		0	/* muhenkan */
#define	RK_XFER		1	/* hiragana henkan */
#define	RK_HFER		2	/* hankaku henkan */
#define	RK_KFER		3	/* katakana henkan */
#define	RK_ZFER		4	/* zenkaku  henkan */

#define	RK_CTRLHENKAN		0xf
#define	RK_HENKANMODE(flags)	(((flags)<<RK_XFERBITS)|RK_CTRLHENKAN)

#define RK_TANBUN		0x01
#define RK_MAKE_WORD		0x02
#define RK_MAKE_EISUUJI		0x04
#define RK_MAKE_KANSUUJI	0x08

/* RkRxDic
 *	romaji/kana henkan jisho
 */
struct RkRxDic	{
    int                 dic;		/* dictionary version: see below */
    unsigned char	*nr_string;	/* romaji/kana taiou hyou */
    int			nr_strsz;	/* taiou hyou no size */
    unsigned char	**nr_keyaddr;	/* romaji key no kaishi iti */
    int			nr_nkey;	/* romaji/kana taiou suu */
    unsigned char       *nr_bchars;     /* backtrack no trigger moji */
    unsigned char       *nr_brules;     /* backtrack no kanouseino aru rule */
};

#define RX_KPDIC 0 /* new format dictionary */
#define RX_RXDIC 1 /* old format dictionary */
#define RX_PTDIC 2 /* large format dictionary (almost equal to KPDIC) */

/* kanakanji henkan */

/* romaji hennkan code */
#define	RK_FLUSH	0x8000	/* flush */
#define	RK_SOKON	0x4000	/* sokuon shori */
#define RK_IGNORECASE	0x2000  /* ignore case */

#define	RK_BIN		0
#define	RK_TXT		0x01

#define	RK_MWD	        0
#define	RK_SWD		1
#define	RK_PRE		2
#define	RK_SUC		3

#define KYOUSEI		0x01		/* jisho_overwrite_mode */

#define	Rk_MWD		0x80		/* jiritsugo_jisho */
#define	Rk_SWD		0x40		/* fuzokugo_jisho */
#define	Rk_PRE		0x20		/* settougo_jisho */
#define	Rk_SUC		0x10		/* setsubigo_jisho */

/* permission for RkwChmod() */
#define RK_ENABLE_READ   0x01
#define RK_DISABLE_READ  0x02
#define RK_ENABLE_WRITE  0x04
#define RK_DISABLE_WRITE 0x08
/* chmod for directories */
#define RK_USR_DIR       0x3000
#define RK_GRP_DIR       0x1000
#define RK_SYS_DIR       0x2000
#define RK_DIRECTORY     (RK_USR_DIR | RK_GRP_DIR | RK_SYS_DIR)
/* chmod for dictionaries */
#define RK_USR_DIC       0	/* specify user dic */
#define RK_GRP_DIC       0x4000	/* specify group dic */
#define RK_SYS_DIC       0x8000	/* specify system dic */

#define PL_DIC		 0x0100
#define PL_ALLOW	 0x0200
#define PL_INHIBIT	 0x0400
#define PL_FORCE	 0x0800

#define	NOENT	-2	/* No such file or directory		*/
#define	IO	-5	/* I/O error				*/
#define	NOTALC	-6	/* Cann't alloc. 			*/
#define	BADF	-9	/* irregal argument			*/
#define	BADDR	-10	/* irregal dics.dir	 		*/
#define	ACCES	-13	/* Permission denied 			*/
#define	NOMOUNT	-15	/* cannot mount				*/
#define	MOUNT	-16	/* file already mounted			*/
#define	EXIST	-17	/* file already exits			*/
#define	INVAL	-22	/* irregal argument			*/
#define	TXTBSY	-26	/* text file busy			*/
#define BADARG	-99	/* Bad Argment				*/
#define BADCONT -100	/* Bad Context				*/
#define OLDSRV    -110
#define NOTUXSRV  -111
#define NOTOWNSRV -112

/* kanakanji henkan */

#ifdef __cplusplus
extern "C" {
#endif

#ifndef CANNAWC_DEFINED
  #define CANNAWC_DEFINED
  // 'cannawc' は、EUC-JP を1文字 = 1ワイド文字にしたもの。
  // 32bit widechar は、元々正常に動いていない。16bit のみとする。
  // See man 5 euc.
  //   G0 ISO 646 IRV              0x0000 - 0x007f.
  //   G1 JIS X 0208 Kanji         the values having the bits 0x8080 set.
  //   G2 JIS X 0201 Hankaku-kana. 0x0080 - 0x00ff.
  //   G3 JIS X 0212 Hojo-Kanji.   0x8000 - 0xff7f excluding the values which
  //                             have the 0x0080 bit set.
  #ifndef CANNA_NEW_WCHAR_AWARE
    #define CANNA_NEW_WCHAR_AWARE
  #endif
  #define CANNA_WCHAR16
typedef uint16_t cannawc;
#endif // !CANNAWC_DEFINED

canna_export(void) RkwFinalize pro((void));
canna_export(int) RkwInitialize pro((const char *));
canna_export(int) RkwCreateContext pro((void));
canna_export(int) RkwCloseContext pro((int));
canna_export(int) RkwDuplicateContext pro((int));
canna_export(int) RkwSetDicPath pro((int, const char *));
canna_export(int) RkwGetDirList pro((int, char *,int));
canna_export(int) RkwGetDicList pro((int, char *,int));
canna_export(int) RkwMountDic pro((int, const char* name, int));
canna_export(int) RkwUnmountDic pro((int, char *));
canna_export(int) RkwRemountDic pro((int, char *, int));
canna_export(int) RkwSync pro((int, const char *));
canna_export(int) RkwGetMountList pro((int, char *, int));
canna_export(int) RkwDefineDic pro((int, const char* dicname,
                                    const cannawc* wordrec));
canna_export(int) RkwDeleteDic pro((int, char *, cannawc *));
canna_export(int) RkwBgnBun pro((int, const cannawc* yomi, int, int));
canna_export(int) RkwEndBun pro((int, int));
canna_export(int) RkwGoTo pro((int, int));
canna_export(int) RkwLeft pro((int));
canna_export(int) RkwRight pro((int));
canna_export(int) RkwXfer pro((int, int));
canna_export(int) RkwNfer pro((int));
canna_export(int) RkwNext pro((int));
canna_export(int) RkwPrev pro((int));
canna_export(int) RkwResize pro((int, int));
canna_export(int) RkwEnlarge pro((int));
canna_export(int) RkwShorten pro((int));
canna_export(int) RkwSubstYomi pro((int, int, int, cannawc *, int));
canna_export(int) RkwStoreYomi pro((int, const cannawc* yomi, int));
canna_export(int) RkwGetLastYomi pro((int, cannawc *, int));
canna_export(int) RkwFlushYomi pro((int));
canna_export(int) RkwRemoveBun pro((int, int));
canna_export(int) RkwGetStat pro((int, RkStat *));
canna_export(int) RkwGetYomi pro((int, cannawc *, int));
canna_export(int) RkwGetHinshi pro((int, cannawc *, int));
canna_export(int) RkwGetKanji pro((int, cannawc *, int));
canna_export(int) RkwGetKanjiList pro((int, cannawc *, int));
canna_export(int) RkwGetLex pro((int, RkLex *, int));
canna_export(int) RkwCvtHira pro((cannawc *, int, const cannawc *, int));
canna_export(int) RkwCvtKana pro((cannawc *, int, const cannawc *, int));
canna_export(int) RkwCvtHan pro((cannawc *, int, const cannawc *, int));
canna_export(int) RkwCvtZen pro((cannawc *, int, const cannawc *, int));
canna_export(int) RkwCvtEuc pro((cannawc *, int, const cannawc *, int));
canna_export(int) RkwCreateDic pro((int, char *, int));
canna_export(int) RkwQueryDic pro((int, const char *, char *, struct DicInfo *));
canna_export(void) RkwCloseRoma pro((struct RkRxDic *));
canna_export(struct) RkRxDic * RkwOpenRoma pro((char *));
canna_export(int) RkwSetUserInfo pro((char *, char *, char *));
canna_export(char *) RkwGetServerName pro((void));
canna_export(int) RkwGetServerVersion pro((int *, int *));
canna_export(int) RkwListDic pro((int, char *, char *, int));
canna_export(int) RkwCopyDic pro((int, char *, char *, char *, int));
canna_export(int) RkwRemoveDic pro((int, char *, int));
canna_export(int) RkwRenameDic pro((int, char *, char *, int));
canna_export(int) RkwChmodDic pro((int, char *, int));
canna_export(int) RkwGetWordTextDic pro((int, const char* dirname,
					 const char* dicname, cannawc *, int));
canna_export(int) RkwGetSimpleKanji pro((int, char *, cannawc *, int,
					 cannawc *, int, unsigned char*, int));
canna_export(int) RkwStoreRange pro((int, cannawc *, int));


void	RkFinalize pro((void));
int     RkInitialize pro((char *));
int    	RkCreateContext pro((void));
int     RkCloseContext pro((int));
int	RkDuplicateContext pro((int));
int	RkSetDicPath pro((int, char *));
int	RkGetDirList pro((int, char *,int));
int	RkGetDicList pro((int, char *,int));
int	RkMountDic pro((int, char *, int));
int	RkUnmountDic pro((int, char *));
int	RkRemountDic pro((int, char *, int));
int	RkSync pro((int, char *));
int	RkGetMountList pro((int, char *, int));
int	RkDefineDic pro((int, char *, unsigned char *));
int	RkDeleteDic pro((int, char *, unsigned char *));
int	RkBgnBun pro((int, unsigned char *, int, int));
int	RkEndBun pro((int, int));
int	RkGoTo pro((int, int));
int	RkLeft pro((int));
int	RkRight pro((int));
int	RkXfer pro((int, int));
int	RkNfer pro((int));
int	RkNext pro((int));
int	RkPrev pro((int));
int	RkResize pro((int, int));
int	RkEnlarge pro((int));
int	RkShorten pro((int));
int	RkSubstYomi pro((int, int, int, unsigned char *, int));
int	RkStoreYomi pro((int, unsigned char *, int));
int	RkGetLastYomi pro((int, unsigned char *, int));
int	RkFlushYomi pro((int));
int	RkRemoveBun pro((int, int));
int	RkGetStat pro((int, RkStat *));
int	RkGetYomi pro((int, unsigned char *, int));
int	RkGetHinshi pro((int, unsigned char *, int));
int	RkGetKanji pro((int, unsigned char *, int));
int	RkGetKanjiList pro((int, unsigned char *, int));
int	RkGetLex pro((int, RkLex *, int));
int	RkCvtHira pro((unsigned char *, int, const unsigned char* src, int));
int	RkCvtKana pro((unsigned char *, int, const unsigned char *, int));
int	RkCvtHan pro((unsigned char *, int, const unsigned char *, int));
int	RkCvtZen pro((unsigned char *, int, const unsigned char *, int));
int	RkCvtNone pro((unsigned char *, int, const unsigned char *, int));
int	RkCvtEuc pro((unsigned char *, int, const unsigned char* src, int));
int	RkQueryDic pro((int, char *, char *, struct DicInfo *));

#ifdef __cplusplus
}
#endif


#ifdef CANNA_EXP_PREDEFINED
#undef CANNA_EXP_PREDEFINED
#else
#undef canna_export
#endif


#endif	/* _RK_h */
/* don't add stuff after this line */
